src/hvmc_net/pre_reduce.rs

```diff
@@ -1,117 +1,146 @@
 // Reduce the compiled networks, solving any annihilations and commutations.
 // This is a useful optimization on its own, but also required by an hvm-core optimization.
 
-use crate::{expand, net_from_runtime, net_to_runtime, rdex, reduce, runtime_net_to_runtime_def};
-use hvmc::{
-  ast::{book_from_runtime, name_to_val, Book},
-  run::{Net, Ptr, OP2, REF},
-};
+use std::collections::HashSet;
+
+use chumsky::container::Container;
+use hvmc::{ast::{Book, Net, Tree}, dispatch_dyn_net, host::Host, run::Tag};
+
+use crate::CORE_BUILTINS;
 
 const MAX_ITERS: usize = 100_000;
 
 /// Reduces the definitions in the book individually, except for main.
 /// If cross_refs, will deref and try to find the smallest net.
 /// Otherwise, just apply node~node interactions.
+///
+/// It does not reduce terms that include builtin definitions
```

Comment:
What is the problem of reducing terms with builtins as long as we don't reduce the builtins themselves?
Let's say for example:
```
(print_pi) =
  (HVM.log <some complicated term that reduces to a constant> *)
```